nvflare/fuel/utils/pipe/pipe_handler.py

```diff
@@ -149,53 +223,70 @@ def _make_event_message(topic: str, data):
     def wait_for_other_end(self, timeout: Optional[float] = None):
         self._other_end_is_up.wait(timeout)
 
-    def send_to_peer(self, msg: Message, timeout=None) -> bool:
+    def send_to_peer(self, msg: Message, timeout=None, abort_signal: Signal = None) -> bool:
         """Sends a message to peer.
 
         Args:
             msg: message to be sent
             timeout: how long to wait for the peer to read the data.
                 If not specified, return False immediately.
+            abort_signal:
 
         Returns:
             Whether the peer has read the data.
         """
         if timeout is not None:
             check_positive_number("timeout", timeout)
         try:
-            return self._send_to_pipe(msg, timeout)
+            return self._send_to_pipe(msg, timeout, abort_signal)
         except BrokenPipeError:
             self._add_message(self._make_event_message(Topic.PEER_GONE, "send failed"))
             return False
 
     def notify_end(self, data):
         """Notifies the peer that the communication is ended normally."""
-        self.send_to_peer(self._make_event_message(Topic.END, data))
+        p = self.pipe
+        if p:
+            try:
+                p.send(self._make_event_message(Topic.END, data))
+            except Exception as ex:
+                self.logger.debug(f"exception notify_end: {secure_format_exception(ex)}")
 
     def notify_abort(self, data):
         """Notifies the peer that the communication is aborted."""
-        self.send_to_peer(self._make_event_message(Topic.ABORT, data))
+        p = self.pipe
+        if p:
+            try:
+                p.send(self._make_event_message(Topic.ABORT, data))
+            except Exception as ex:
+                self.logger.debug(f"exception notify_abort: {secure_format_exception(ex)}")
 
     def _add_message(self, msg: Message):
         if msg.topic in [Topic.END, Topic.ABORT, Topic.PEER_GONE]:
             if self.status_cb is not None:
                 self.status_cb(msg, *self.cb_args, **self.cb_kwargs)
                 return
+        else:
+            if self.msg_cb is not None:
+                self.msg_cb(msg, *self.msg_cb_args, **self.msg_cb_kwargs)
+                return
+
         with self.lock:
             self.messages.append(msg)
 
     def _read(self):
         try:
             self._try_read()
         except Exception as e:
+            self.logger.error(f"read error: {secure_format_exception(e)}")
             self._add_message(self._make_event_message(Topic.PEER_GONE, f"read error: {secure_format_exception(e)}"))
 
     def _try_read(self):
         last_heartbeat_received_time = time.time()
         last_heartbeat_sent_time = 0.0
         while not self.asked_to_stop:
             now = time.time()
-            msg = self._receive_from_pipe()
+            msg = self.pipe.receive()
             if msg:
                 last_heartbeat_received_time = now
                 self._other_end_is_up.set()
```

Comment:
This event is used to indicate that the other end (the peer) is up. You need to check the msg's topic and decide whether this is truly a message from the other end.  Only if the topic is not one of (PEER_GONE, ABORT, END), you can say the peer is up.

Reply:
good point, fixed.